src/hyperplane/aws_s3_utils.py

```diff
@@ -0,0 +1,200 @@
+            target_name = relpath_to_source
+        elif target_path.endswith('/'):
+            # target path is a dir, keep original files` hierarchy.
+            target_name = target_path + relpath_to_source
+        else:
+            # uploading file to a specific target
+            target_name = target_path
+
+        dir_to = os.path.dirname(target_name)
+        Path(dir_to).mkdir(parents=True, exist_ok=True)
+
+        bucket.download_file(obj.key, target_name)
+
+
+def upload_to_s3(source_path: str, target_path: str = None, regex_filter: str = None, verbose: bool = True):
+    """
+    Upload source path to {bucket}/{target_path}
+    Args:
+        source_path: if a directory, all files and directories under source_path will be upload recursively under target_path
+        target_path: full target path including desired file name for files and directory name for directory.
+        regex_filter: if not None, only filename that matches this pattern will be uploaded
+        bucket_name: bucket name to upload
+        verbose: show prints
+    Returns:
+        None
+    """
+    user_s3_creds = get_s3_credentials()
+    bucket_name = user_s3_creds.get("bucket_url")
+    if not bucket_name:
+        report("User didn't specify bucket name and have no bucket configurations in our system")
+        return None
+    s3_client = _connect_to_s3_bucket(bucket_name=bucket_name, user_s3_creds=user_s3_creds,
+                                      access_key_id=user_s3_creds.get("access_key_id"),
+                                      access_key_secret=user_s3_creds.get("secret_access_key"))
+    if os.path.isfile(source_path):
+        if verbose:
+            print('uploading file to s3')
+        return _upload_file_to_s3(file_path=source_path, bucket=bucket_name, object_name=target_path,
+                                  regex_filter=regex_filter, s3_client=s3_client, verbose=verbose)
+    elif os.path.isdir(source_path):
+        if verbose:
+            print('uploading dir to s3')
+        return _upload_directory_to_s3(local_base_directory=source_path, bucket=bucket_name,
+                                       s3_base_directory=target_path,
+                                       regex_filter=regex_filter, s3_client=s3_client, verbose=verbose)
+
+
+def _upload_file_to_s3(file_path, bucket, s3_client, object_name=None, extra_args=None, verbose=False,
+                       regex_filter: str = None):
+    """Upload a file to an S3 bucket
+
+    :param file_path: File to upload
+    :param bucket: Bucket to upload to
+    :param object_name: S3 object name. If not specified then file_name is used
+    :param extra_args: arg to pass to s3 call
+    :param verbose: print logs if true
+    :param regex_filter: regex compile str to match
+    :param s3_client: s3 client to interact with the bucket
+    :return: s3 url if file was uploaded, else None
+
+    """
+    if regex_filter:
+        regex_filter = re.compile(regex_filter)
```

Comment:
compile this outside of this function, it makes no sense to compile every time for every file (there could be 100o's of them)